source/extensions/filters/http/adaptive_concurrency/controller/gradient_controller.cc

```diff
@@ -215,10 +215,12 @@ void GradientController::recordLatencySample(MonotonicTime rq_send_time) {
     sample_count = hist_sample_count(latency_sample_hist_.get());
   }
 
-  if (inMinRTTSamplingWindow() && sample_count >= config_.minRTTAggregateRequestCount()) {
+  const auto agg_count = config_.minRTTAggregateRequestCount();
+  if (min_rtt_update_mtx_.TryLock() && inMinRTTSamplingWindow() && sample_count >= agg_count) {
```

Comment:
Looks like we already have the mutex lock for most logics in updateMinRTT(): https://github.com/envoyproxy/envoy/blob/aa73d170ffdc67a160a675a3c22f66445e5badc5/source/extensions/filters/http/adaptive_concurrency/controller/gradient_controller.cc#L115,  is `min_rtt_update_mtx_ ` just for avoiding the duplicate [schedules for min_rtt_calc_timer_ and sample_reset_timer_](https://github.com/envoyproxy/envoy/blob/aa73d170ffdc67a160a675a3c22f66445e5badc5/source/extensions/filters/http/adaptive_concurrency/controller/gradient_controller.cc#L124-L126) when concurrent requests in different threads flow in. But i am a little fuzzy about how the duplicate schedules for min_rtt_calc_timer_ and sample_reset_timer_ would cause this min_rt issue, could you elaborate?

Reply:
This is for a situation where 2 threads reach this conditional at the same time. If that happens, you have one thread updating the minrtt while the other blocks on the sample mutation mutex. When the minrtt gets updated, the other thread will then update the minrtt but have no samples in the histogram.

Reply:
Thanks for the explanation, it makes senses. This patch will enforce updateMinRTT() to get invoked sequentially. 

But the below event sequence seems to cause the issue again: (assume the event sequences are same as the ones causing the issue, except for having enterMinRTTSamplingWindow() between two threads' updateMinRTT())
```
updateMinRTT() in work thread 1 -> set deferred_limit_value_ to be zero, and clear histogram
enterMinRTTSamplingWindow() in any thread-> set deferred_limit_value_ to be non zero.
updateMinRTT() in work thread 2 -> do the percentile calculation on the empty histogram, which causes the issue.
```
 And my test in https://github.com/envoyproxy/envoy/pull/15001 seems similar to this case.

Reply:
If I understand the scenario correctly, I don't think this can happen because of the TryLock() being a non-blocking call. If another thread holds the mutex, the if-statement will always evaluate to false on other threads and you can't call `updateMinRTT()` from `recordLatencySample()`.

Also, `enterMinRTTSamplingWindow()` has a conditional to verify we're not in a minRTT measurement window?

Let me know if I'm misunderstanding you. I want this race to be fully fixed.

Reply:
Understood TryLock() is a non-blocking call, and we never have two simultaneous updateMinRTT() based on the logic. But for the even sequence in my above example, i assumed it is caused by time organically, not about updateMinRTT() of thread 2 pends for for the updateMinRTT() of thread 1 due to the mutex lock. 

Taking a more concrete example here: (assume config_.minRTTAggregateRequestCount() is 1, and the condition `sample_count >= config_.minRTTAggregateRequestCount()` in recordLatencySample() is almost alway true. In this case we can focus on the condition  `min_rtt_update_mtx_TryLock()` and `inMinRTTSamplingWindow()`)

- Thread 1 reaches  min_rtt_update_mtx_.TryLock() at 00:00s , acquire the lock and invoke updateMinRTT(), then release the lock at 00:01s.   -> deferred_limit_value_ is set to be zero, which means inMinRTTSamplingWindow() become false.
- enterMinRTTSamplingWindow() gets invoked at 00:02s, proceed to execute as inMinRTTSamplingWindow() is false: set deferred_limit_value_ to be non-zero, and  clear the histogram, exit at 00:03s. -> deferred_limit_value_ is set to be non-zero, which means inMinRTTSamplingWindow() becomes true.
- Thread 2 reaches min_rtt_update_mtx_.TryLock() at 00:04s, acquire the lock and proceed to invoke updateMinRTT() as inMinRTTSamplingWindow() is true, then do the percentile calculation on the empty histogram.



> enterMinRTTSamplingWindow() has a conditional to verify we're not in a minRTT measurement window?

updateMinRTT() in work thread 1 sets deferred_limit_value_ to be 0, which cause inMinRTTSamplingWindow() in enterMinRTTSamplingWindow() becomes false.



Not sure how realistic my example is, but let me know if it doesn't sound right.

Reply:
I just updated the test in https://github.com/envoyproxy/envoy/pull/15001/files#diff-e4b3862c87279e6e5c6253e8f126985dd18a3750546b9d43559fd088b9ce32dfR397-R403, to make it more similar to the real usage, except for directly invoking updateMinRTT(). And it also caused the extremely large minRTT issue.
```
   tryForward(controller, true); // thread 1
  sampleLatency(controller, std::chrono::milliseconds(1)); // thread 1
  tryForward(controller, true); // thread 2
  sampleLatency(controller, std::chrono::milliseconds(1)); // thread 2
  updateMinRTT(controller); // thread 1
  enterMinRTTSamplingWindow(controller); // any thread
  updateMinRTT(controller); // thread 2
```